File: 0931-minimum-falling-path-sum/0931-minimum-falling-path-sum.go
func minFallingPathSum(matrix [][]int) int {
    n := len(matrix)
    dp := make([][]int, n)
    for i:=0;i<n;i++ {
        dp[i] = make([]int, n)
        for ii:=0;ii<n;ii++{
            dp[i][ii] = math.MaxInt
        }
    }

    moves := [][]int {
        {-1, 0},
        {-1, -1},
        {-1, 1},
    }
    for i:=0;i<n;i++{
        for ii:=0;ii<n;ii++{
            if i == 0 {
                dp[i][ii] = matrix[i][ii]
            }else{
                for _, move := range moves {
                    r, c := i+move[0], ii+move[1]
                    if r >= 0 && r < n && c >= 0 && c < n {
                        dp[i][ii] = min(dp[i][ii], matrix[i][ii] + dp[r][c])
                    }
                }
            }
        }
    }
    res := dp[n-1][0]
    for i:=1;i<n;i++{
        res = min(res, dp[n-1][i])
    }
    return res
}

File: 0931-minimum-falling-path-sum/0931-minimum-falling-path-sum.rs
use std::cmp::{max, min};

impl Solution {
    pub fn min_falling_path_sum(matrix: Vec<Vec<i32>>) -> i32 {
        if matrix.len() == 1 {
            return *matrix[0].iter().min().unwrap();
        }
        let mut prev_sums = matrix[0].clone();
        for r in 1..matrix.len() {
            let mut curr_sums = vec![0; prev_sums.len()];
            for (c, &grid_val) in matrix[r].iter().enumerate() {
                let left = max(0, c as i32 - 1) as usize;
                let right = min(curr_sums.len() - 1, c + 1);

                curr_sums[c] = prev_sums[left..right + 1].iter().min().unwrap() + grid_val;
            }
            prev_sums = curr_sums;
        }
        return prev_sums.into_iter().min().unwrap();
    }
}